### PR TITLE
Fix zero-dim handling in torch.matmul

### LINKED ISSUE #63352: torch.matmul doesn't work on certain tensors with zero-sized dimensions
## 🐛 Bug

```
import torch
x = torch.randn(3, 0, 0)
y = torch.randn(0, 0)
x @ y
```
this gives the following error:
```
RuntimeError: cannot reshape tensor of 0 elements into shape [-1, 0] because the unspecified dimension size -1 can be any value and is ambiguous
```

## Expected behavior

Returns a tensor of shape (3, 0, 0)

## Environment

Master

## Additional context

Discovered while doing OpInfo tests for functorch

cc @jianyuh @nikitaved @pearu @mruberry @heitorschueroff @walterddr @IvanYashchuk @xwang233 @Lezcano

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#63359 Fix zero-dim handling in torch.matmul**

Fixes #63352. The problem was that in e.g. `torch.matmul(A, B)` with A,
B having shapes [3, 2, 0] and [0, 2], the code attempts to call
`A.view(-1, 0)` which fails due to "-1 being ambiguous". The solution is
to manually compute what we want the shape of the view to be.

Test Plan:
- new tests

Differential Revision: [D30351583](https://our.internmc.facebook.com/intern/diff/D30351583)

### CODE REVIEW

**Comment by zou3519** [comment]
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30351583).

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 1]
```diff
@@ -1455,7 +1455,15 @@ Tensor matmul(
     }
 
     // fold the batch into the first dimension
-    Tensor t1 = tensor1.expect_contiguous()->view({-1, size1[size1.size() - 1]});
+    // Why not tensor1.view(-1, size1[size1.size() -1])?
+    // If the last dim is 0, then view(-1, 0) won't work because the -1 becomes ambiguous.
+    // This can happen in e.g. [3, 5, 0] @ [0, 0].
+    // So we manually compute the folding as a result.
+    auto dim1_size = 1;
+    for (const auto idx : c10::irange(0, size1.size() - 1)) {
```
Comment (ngimel) [inline aten/src/ATen/native/LinearAlgebra.cpp:1463]: use c10::multiply_integers instead of this loop

**Review by IvanYashchuk — APPROVED** [review]

**Review by lezcano — APPROVED** [review]
If this change adds instructions to matmul (I don't know if it does), it might be interesting to just check `if C10_UNLIKELY(numel() == 0)` at the beginning and handle this case separately.

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 1 above)
Comment (zou3519) [reply aten/src/ATen/native/LinearAlgebra.cpp:1463]: Thanks for the pointer, I didn't know this existed.

**Comment by zou3519** [comment]
@Lezcano thanks for the suggestion to benchmark instruction counts. It looks like this PR adds one additional instruction to the non-zero-dim case https://gist.github.com/zou3519/e56d2619ddb3de5473c194f20f82c153 so it's not worth having a branch (the branch will add more instructions than it saves).

**Comment by zou3519** [comment]
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30351583).

**Pull request closed without merging**